File: net/JetBrains.Serialization/src/Dataholders/X500NameInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Org.BouncyCastle.Asn1.X500;
using Org.BouncyCastle.Asn1.X509;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class X500NameInfo : IEncodableInfo
{
  public string Name { get; }
  public List<List<RdNInfo>> RdNs { get; }

  public X500NameInfo(string name, List<List<RdNInfo>> rdNs)
  {
    Name = name;
    RdNs = rdNs;
  }

  public X500NameInfo(X509Name issuer)
    : this(issuer.ToString(),
      ((DerSequence)issuer.ToAsn1Object()).ToArray().OfType<DerSet>().Select(set =>
        set.ToArray().OfType<DerSequence>().Select(rdn =>
          new RdNInfo(TextualInfo.GetInstance(rdn[0]), TextualInfo.GetInstance(rdn[1]))).ToList()).ToList()
    )
  {
  }

  private DerSequence ToDLSequence()
  {
    return
      RdNs.Select(
        rdnList => rdnList.Cast<IEncodableInfo?>().ToList().ToPrimitiveList().ToDerSet()
      ).Cast<Asn1Encodable>().ToList()!.ToDerSequence();
  }

  public Asn1Encodable ToPrimitive() => ToDLSequence();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/X500NameInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.ASN1Primitive
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.x500.X500Name

@Serializable
data class X500NameInfo(
  val name: String,
  val rdNs: List<List<RdNInfo>>
) : EncodableInfo {
  constructor(issuer: X500Name) : this(issuer.toString(),
    issuer.rdNs.map {
      it.typesAndValues.map { tv ->
        RdNInfo(
          TextualInfo.getInstance(tv.type),
          TextualInfo.getInstance(tv.value)
        )
      }
    })

  private fun toDLSequence(): DLSequence =
    rdNs.map {
      it.map { info ->
        info.toPrimitive()
      }.toDLSet()
    }.toDLSequence()

  override fun toPrimitive(): ASN1Primitive = toDLSequence().toASN1Primitive()
}